corehq/motech/repeaters/_design/views/repeat_records/map.js

```diff
@@ -10,5 +10,7 @@ function (doc) {
         }
         emit([doc.domain, doc.repeater_id, state, doc.last_checked], null);
         emit([doc.domain, null, state, doc.last_checked], null);
+        emit([doc.domain, null, null, doc.last_checked], null);
```

Comment:
I am wondering if I should preindex the Prod. There seems to be about 57million repeat-record docs on Prod.

Reply:
Probably a good idea. You'll also need to run `makemigrations` to address the test failure.
```
[test-django-migrations.sh] ERROR: Migrations are missing.  Did you run './manage.py makemigrations'?
```

Reply:
While the reindex is in progress (from setup-release), how do I make sure that the (not-yet-updated) queries continue to work correctly ? The reindex I made with old queries will result in duplicates.

Reply:
It sounds like you are saying that the results will not be backward-compatible. This will need to be revised. All view changes must be backward compatible for a safe rollout.

Reply:
In that case perhaps it would be better to create a new view in couchapps rather than change this one.

Reply:
Ah, that's a good idea. Since it seems rather impossible to write a query that's backwards compatible, I will create a new view and drop the old view in a separate PR.

Reply:
Or I guess, I can drop it in the same PR as well?

Reply:
@millerdev Regarding the migrations, I updated `couchviews.lock` file. Does the preindex util also drop the old view index or should that be done manually after?

Reply:
> Or I guess, I can drop it in the same PR as well?

Dropping the view does not actually remove anything from Couch, right? If that's true, then this is fine since previously deployed code will continue to access the view while this is being deployed. If dropping it makes the view inaccessible to running code then it should be dropped in a separate PR and deployed at least 6 weeks later.

> Does the preindex util also drop the old view index or should that be done manually after?

It should be done manually, probably at least 6 weeks after this is merged. @dannyroberts has done this recently for other views, so he is probably the best person to ask about that process.

Reply:
@sravfeyn I think you misunderstood what I was suggestion. I wasn't suggesting replacing this view with a new one. I mean to create a new view with just the output you need for the filter query. Then you can leave the existing view as it is and don't have to change all the references to it.

Reply:
The issue with that is, that the new query will have to query both the views and combine the results. The existing index can't be queried on a specific combination of the keys, so I have to modify it to support that.

Reply:
Since couchdb reindexes all views in a design doc together, dropping (or editing) a view that is in a design doc that has other views in it requires reindexing all those other views as well.

I'll note that this "repeaters" design doc is our very largest remaining design doc in couch, and so making sure the new version of the index doesn't fill up the disk before the old one is deleted is quite important. By a rough estimate, we should not hit that limit, but be sure to watch for it if you go this route.

Reply:
I *think* based on my understanding of the motivation here and the discussion so far, I would advocate for

(1) leaving the "repeaters" design doc, including `repeaters/repeat_records`
(2) creating a new view in `couchapps`, `repeat_records_by_last_check`, that has exactly the contents you currently have in `repeaters/repeat_records_by_last_check` in this PR

And optionally as cleanup steps:
(3) In a different PR, copying `repeat_records_by_last_check` and `repeat_records_by_payload_id` each into their own design doc under `couchapps` and change all code to reference those instead. (`repeaters/repeaters` is now unused, so no need to copy.)
(4) In a final PR, delete repeaters/_design entirely, making sure no currently deployed code uses it.

Feel free to consider it done at this point. There are some manual operational steps one can do to remove the index from disk, but I can make sure those happen afterwards if I want to lower disk usage.

Reply:
@dannyroberts do you think all of this is necessary given that we are in the process of migrating repeat records to SQL?

Reply:
> I'll note that this "repeaters" design doc is our very largest remaining design doc in couch, and so making sure the new version of the index doesn't fill up the disk before the old one is deleted is quite important. By a rough estimate, we should not hit that limit, but be sure to watch for it if you go this route.

Phew, I did not know this! I guess I will go ahead, unless Simon's comment changes anything.

> @dannyroberts do you think all of this is necessary given that we are in the process of migrating repeat records to SQL?

My understanding is that it is still useful since there are about 4 views under repeaters that will get reindexed otherwise and we don't know (?) when we will migrate them to SQL. But I will differ to the result of your discussion.

Reply:
Yeah fair question, Simon. I think it would be totally reasonable to just add the new view and not touch the old ones (i.e. do steps 1 and 2 above) and postpone the decision about any cleanup until we see the impact on disk usage; if it's fairly minor in the scheme of things then just waiting it out seems reasonable. That said like Sravan noted, it's hard to predict exactly when the repeat records migration will be completed, so if it _does_ seem to impact anything (disk usage, performance, etc) then we'd want to address it rather than waiting. But I think there's a good chance it won't be necessary.

Reply:
Thanks Danny.

> I think it would be totally reasonable to just add the new view and not touch the old ones (i.e. do steps 1 and 2 above)

I think we have an agreement on this, I will take this route. FYI @snopoke

Reply:
I was mostly referring to the cleanup steps (3 & 4). I think steps 1 & 2 are good because it is more backwards compatible and easier to rollback if necessary.

Reply:
Okay, thanks. I will proceed with steps 1 and 2